### PR TITLE
feat(suite-common): common connect for suite and suite-native

### LINKED ISSUE #5708: suite-native: use @trezor/connect for watch-only implementation
We will need connect for fetching transactions and fiat prices etc from blockbook. This PR should be just simply adding connect to mobile app, init it and test if we can fetch transactions etc. 

Storing of transaction to redux, rendering them etc. is NOT part of this issue. There will be follow up task (TODO link it here).
Connecting to Trezor device is NOT part of this issue.

**Implementation ideas/questing:**
1. Will connect work without device, for just xpub?
2. I would imagine we could have package, for these connect-setup stuff.3. 
3. Would be really nice if that package could be shared between suite, I think it could be pretty achievable. From what I have seen in suite code these two files will be most interesting for you:
`packages/suite/src/actions/suite/trezorConnectActions.ts`
`packages/connect-explorer/src/reducers/trezorConnectReducer.ts`

4. I think these two files could be moved to some brand new `@suite-common/connect` package (you may need to add suite-common workspace in yarn). Then you can refactor them to be reusable and to use `redux-toolkit` in new package. There will some challenges but is definitely doable, reducer should be simple, actions file will harder because there are some specific actions related suite which we don't have in mobile yet, so we will probably need two different init actions. You can use react native file suffix to achieve that like `initConnect.native.tsx`.

5. Also please pay attention that you **can't import from suite/ in new package** because that would be cyclic dependency, so maybe you will need to move some things to this package or create thrid package that will be used both with connect package and old suite core package. Another approach could be use something like generator function like `const generateInitAction = ({erroAction) => (dispatch, getState) => { dispatch(errorAction(err)) ...... }` where you will generate init function and can pass these specific functions or whatever to generator first.

6. **What I totally recommend is to do TDD**, before you even beggin refactoring, write tests for that old code and then start refactoring and in the end, you should pass this tests also with refactored code. That should minimalize risk of breaking desktop suite during refactor. Test for `actions` are actually here, so verify if they are enough and for `reducer` you have to write them. **Without doing TDD refactoring will take us to bug hell.**7. 

7. You might need to change `createStore` in suite to `configureStore` from redux-toolkit. 

**People that can help you:**
- @mroz22 - I totally recommend to have call with him before begin. Let him do to some introduction of connect, discuss some possibilties with him etc.
- @matejkriz - Suite related stuff

### PR DESCRIPTION
**Pull request by bouchja1**
- the new package **@suite-common/connect-init** for Suite and Suite Native handling Trezor Connect initialization and patches.
- _trezorConnectActions_ in Suite modified to use common init function from the new package
- Suite Native and Trezor Connect init added

Closes https://github.com/trezor/trezor-suite/issues/5708

### CODE REVIEW

File: packages/suite/src/reducers/store.ts [hunk 1]
```diff
@@ -19,19 +18,16 @@ import backupReducers from '@backup-reducers';
 // it's not included into `@suite-middlewares` index
 import toastMiddleware from '@suite-middlewares/toastMiddleware';
 
-const rootReducer = combineReducers({
+const reducer = {
```
Comment (Nodonisko) [inline packages/suite/src/reducers/store.ts]: reducer**s**

File: packages/suite/src/reducers/store.ts [hunk 2]
```diff
@@ -57,13 +52,9 @@ if (!process.env.CODESIGN_BUILD) {
     middlewares.push(logger);
 }
 
-/* eslint-disable no-underscore-dangle */
-const composeEnhancers =
-    typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
```
Comment (Nodonisko) [inline packages/suite/src/reducers/store.ts]: Don't forget to bring this blacklist back

File: suite-common/connect/package.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "@suite-common/connect",
```
Comment (Nodonisko) [inline suite-common/connect/package.json]: I don't like name connect for this package. It will be confusing, we should think about something moreover different from universal connect packages. Maybe somethings like connect-init but better suggestions welcomed.

File: suite-common/connect/src/actions/cardanoPatch.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import TrezorConnect from '@trezor/connect';
+import { createAsyncThunk } from '@reduxjs/toolkit';
+
+export const cardanoPatch = createAsyncThunk<
+    void,
+    void,
+    {
+        state: any;
+    }
+>('cardanoPatch', (_, { getState }) => {
```
Comment (Nodonisko) [inline suite-common/connect/src/actions/cardanoPatch.ts]: Not sure if this needs to be thunk. We can make it function and just pass `enabledNetworks` as param.

File: suite-common/connect/src/actions/cardanoPatch.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import TrezorConnect from '@trezor/connect';
+import { createAsyncThunk } from '@reduxjs/toolkit';
+
+export const cardanoPatch = createAsyncThunk<
```
Comment (Nodonisko) [inline suite-common/connect/src/actions/cardanoPatch.ts]: I would suggest better name like `cardanoConnectPatch` to make more obvious that this is patching connect lib

File: suite-common/connect/src/actions/eventListeners.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import TrezorConnect, {
+    BLOCKCHAIN_EVENT,
+    DEVICE_EVENT,
+    TRANSPORT_EVENT,
+    UI_EVENT,
+} from '@trezor/connect';
+import { createAsyncThunk, Dispatch } from '@reduxjs/toolkit';
+
```
Comment (Nodonisko) [inline suite-common/connect/src/actions/eventListeners.ts]: bindConnectActionsListeners

File: suite-common/connect/src/actions/wrapMethods.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import TrezorConnect from '@trezor/connect';
+import { AnyAction, createAsyncThunk, Dispatch } from '@reduxjs/toolkit';
+
```
Comment (Nodonisko) [inline suite-common/connect/src/actions/wrapMethods.ts]: lockDeviceConnectPatch

File: suite-common/connect/src/actions/wrapMethods.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+export const wrapMethods = createAsyncThunk<
+    void,
+    {
+        lockDevice: (payload: boolean) => AnyAction;
+    },
+    {
+        dispatch: Dispatch;
+    }
+>('wrapMethods', ({ lockDevice }, { dispatch }) => {
+    const wrappedMethods = [
```
Comment (Nodonisko) [inline suite-common/connect/src/actions/wrapMethods.ts]: Put this const outside

File: suite-common/connect/src/actions/wrapMethods.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        'cardanoSignTransaction',
+        'backupDevice',
+        'recoveryDevice',
+    ] as const;
+    wrappedMethods.forEach(key => {
+        // typescript complains about params and return type, need to be "any"
+        const original: any = TrezorConnect[key];
+        if (!original) return;
+        (TrezorConnect[key] as any) = async (params: any) => {
+            dispatch(lockDevice(true));
```
Comment (Nodonisko) [inline suite-common/connect/src/actions/wrapMethods.ts]: Maybe we should make all these patches redux agnostic. What do you think?

 if we make this just function then we can do something like 

```tsx
const lockDeviceConnectPatch = ({onDeviceLock, onDeviceUnlock}) => {
...
onDeviceLock();
const result = await original(params);
onDeviceUnlock();
...

// usage in init action
lockDeviceConnectPatch({
  onDeviceLock: () => dispatch({lockDevice(true)}),
  onDeviceUnlock: () => dispatch({lockDevice(false)}),
})
```

File: suite-common/connect/src/actions/initConnect.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
+import TrezorConnect from '@trezor/connect';
+import { AnyAction, createAsyncThunk, Dispatch } from '@reduxjs/toolkit';
+import { cardanoPatch } from './cardanoPatch';
+import { resolveStaticPath } from '@trezor/utils';
+import { eventListeners } from './eventListeners';
+import { wrapMethods } from './wrapMethods';
+
+export const prepareInit = ({
```
Comment (Nodonisko) [inline suite-common/connect/src/actions/initConnect.ts]: `prepareConnectInit` and file should be named same

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/connect/package.json (hunk 3 above)
Comment (bouchja1) [reply suite-common/connect/package.json]: connect-init is ok

File: suite-common/connect/src/actions/cardanoPatch.ts (hunk 4 above)
Comment (bouchja1) [reply suite-common/connect/src/actions/cardanoPatch.ts]: https://github.com/trezor/trezor-suite/pull/5760/commits/aca99f3d8491e5725750f30a77cee03823be443d#diff-08a808ab26ecc83646b9ef87e0db26ad73ef35f40c6af8fa3aa59398a760a898

File: suite-common/connect/src/actions/wrapMethods.ts (hunk 8 above)
Comment (bouchja1) [reply suite-common/connect/src/actions/wrapMethods.ts]: https://github.com/trezor/trezor-suite/pull/5760/commits/aca99f3d8491e5725750f30a77cee03823be443d#diff-6fa62ef2e598fb948540f5614ade60317bf8429f25acd54a8da018c9ea0b50f5

File: suite-common/connect/src/actions/wrapMethods.ts (hunk 9 above)
Comment (bouchja1) [reply suite-common/connect/src/actions/wrapMethods.ts]: I agree. It is definitely a more elegant solution! https://github.com/trezor/trezor-suite/pull/5760/commits/aca99f3d8491e5725750f30a77cee03823be443d#diff-f8b4494e69ec35a7431e26e9f0928a1da250bb00abaaebf905e91fa19de47bdc

File: suite-common/connect-init/src/cardanoConnectPatch.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+export const cardanoConnectPatch = (enabledNetworks: string[]) => {
+    // Pass additional parameter `useCardanoDerivation` to Trezor Connect methods
+    // in order to enable cardano derivation on a device
+    // https://github.com/trezor/trezor-firmware/blob/master/core/src/apps/cardano/README.md#seed-derivation-schemes
+    type ConnectKey = keyof typeof TrezorConnect;
+
+    // List of methods that doesn't work with additional `useCardanoDerivation` param
+    // (eg. because they don't accept options object as a param)
+    // or they don't trigger seed derivation on a device so there is no need to pass it.
+    const blacklist: ConnectKey[] = [
```
Comment (Nodonisko) [inline suite-common/connect-init/src/cardanoConnectPatch.ts]: You can keep this in this file but place it outside of function

File: suite-native/app/src/navigation/RootTabNavigator.tsx [hunk 12]
```diff
@@ -12,19 +14,27 @@ import { RootTabsParamList, RouteTabs, rootTabsOptions } from './routes';
...
-        <Tab.Screen name={RouteTabs.Accounts} component={AccountsScreen} />
-        <Tab.Screen name={RouteTabs.Action} component={ActionScreen} />
-        <Tab.Screen name={RouteTabs.Prices} component={PricesScreen} />
-        <Tab.Screen name={RouteTabs.SettingsStack} component={SettingsStackNavigator} />
-    </Tab.Navigator>
-);
+export const RootTabNavigator = () => {
+    const dispatch = useDispatch();
+
+    useEffect(() => {
```
Comment (Nodonisko) [inline suite-native/app/src/navigation/RootTabNavigator.tsx]: Not a right place,this is just for navigation stuff so don't mix it. App.ts will be better place

File: suite-common/connect-init/src/lockDeviceConnectPatch.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+import { TrezorConnectWrappedMethod } from './wrappedConnectMethods';
+import { AnyAction } from '@reduxjs/toolkit';
+
+export const lockDeviceConnectPatch = ({
+    wrappedMethods,
+    onDeviceLock,
+    onDeviceUnlock,
+}: {
+    wrappedMethods: readonly TrezorConnectWrappedMethod[];
+    onDeviceLock: () => AnyAction;
```
Comment (Nodonisko) [inline suite-common/connect-init/src/lockDeviceConnectPatch.ts]: This type is actually () => void

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/connect-init/src/lockDeviceConnectPatch.ts (hunk 13 above)
Comment (bouchja1) [reply suite-common/connect-init/src/lockDeviceConnectPatch.ts]: ah, copy-paste typo

File: suite-native/app/src/navigation/RootTabNavigator.tsx (hunk 12 above)
Comment (bouchja1) [reply suite-native/app/src/navigation/RootTabNavigator.tsx]: This was mainly for demo purposes. I will create some "preloader" component for connect init and put it under redux Provider

File: packages/suite/src/actions/suite/trezorConnectActions.ts [hunk 14]
```diff
@@ -1,154 +1,42 @@
...
-        });
+export const init = createAsyncThunk<
+    void,
+    void,
+    {
+        state: AppState;
     }
-};
+>('initConnect', (_, { dispatch, getState }) => {
+    const { enabledNetworks } = getState().wallet.settings;
```
Comment (bouchja1) [inline packages/suite/src/actions/suite/trezorConnectActions.ts]: @Nodonisko ...or can be your proposed version with selector (sth like `selectWalletSettings(getState()))`. If I am thinking about it now, your proposal seems better to me because if we go in a "selector way" for other future things we will need to pass from suite to common suite, it can be better for encapsulation of some potentially complex lookup logic if there is some later or sth like.

File: suite-native/app/src/navigation/RootTabNavigator.tsx (hunk 12 above)
Comment (bouchja1) [reply suite-native/app/src/navigation/RootTabNavigator.tsx]: https://github.com/trezor/trezor-suite/pull/5760/commits/704afb4bb71800a0363e3852b06cbd9000b9bcc5

File: packages/suite/src/reducers/store.ts (hunk 2 above)
Comment (bouchja1) [reply packages/suite/src/reducers/store.ts]: I returned it back to the version without `configureStore` from **redux-toolkit** because the Suite web app was not able to start and was frozen. jfyi not because of that blacklist (it is possible to pass as an object to a new redux-toolkit _devTools_ prop of configureStore)

File: packages/suite/src/actions/suite/trezorConnectActions.ts (hunk 14 above)
Comment (Nodonisko) [reply packages/suite/src/actions/suite/trezorConnectActions.ts]: Yea I would prefer it selector way

File: suite-common/connect-init/src/actions/prepareConnectInit.ts [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+import { cardanoConnectPatch } from '../cardanoConnectPatch';
+import { wrappedMethods } from '../wrappedConnectMethods';
+import { lockDeviceConnectPatch } from '../lockDeviceConnectPatch';
+import { bindConnectActionsListeners, ConnectEventAction } from '../bindConnectActionsListeners';
+
+export const prepareConnectInit = createAsyncThunk<
+    void,
+    {
+        lockDevice: (payload: boolean) => AnyAction;
+        setInitConnectError: (payload: any) => AnyAction;
```
Comment (Nodonisko) [inline suite-common/connect-init/src/actions/prepareConnectInit.ts]: Can we use something better than any?

File: suite-native/state/src/configureStore.ts [hunk 16]
```diff
@@ -8,3 +8,4 @@ export const store = configureStore({
 });
 
 export type RootState = ReturnType<typeof store.getState>;
+export type AppDispatch = typeof store.dispatch; // https://stackoverflow.com/a/72059282
```
Comment (Nodonisko) [inline suite-native/state/src/configureStore.ts]: Get rid of this, no benefits have this typed

File: packages/suite/src/actions/suite/trezorConnectActions.ts [hunk 17]
```diff
@@ -1,154 +1,35 @@
...
+    // 'https://localhost:8088/';
+    // 'https://connect.corp.sldev.cz/develop/';
+
+    dispatch(
+        prepareConnectInit({
+            lockDevice,
+            setInitConnectError,
+            onConnectInitialized,
+            enabledNetworks,
+            initSettings: {
```
Comment (Nodonisko) [inline packages/suite/src/actions/suite/trezorConnectActions.ts]: I would prefer to have this in const at the top of file

File: suite-common/connect-init/src/actions/prepareConnectInit.ts [hunk 18]
```diff
@@ -0,0 +1,50 @@
+import TrezorConnect, { ConnectSettings, Manifest } from '@trezor/connect';
+import { AnyAction, createAsyncThunk } from '@reduxjs/toolkit';
+import { cardanoConnectPatch } from '../cardanoConnectPatch';
+import { lockDeviceConnectPatch } from '../lockDeviceConnectPatch';
+import { bindConnectActionsListeners, ConnectEventAction } from '../bindConnectActionsListeners';
+
+export const prepareConnectInit = createAsyncThunk<
+    void,
+    {
+        lockDevice: (payload: boolean) => AnyAction;
```
Comment (Nodonisko) [inline suite-common/connect-init/src/actions/prepareConnectInit.ts]: For all action creators please use types `ActionCreatorWithPayload` or `ActionCreatorWithoutPayload` from redux-toolkit. So for example here it will be `lockDevice: ActionCreatorWithPayload<boolean>`

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/connect-init/src/cardanoConnectPatch.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+export const cardanoConnectPatch = (enabledNetworks: string[]) => {
+    // Pass additional parameter `useCardanoDerivation` to Trezor Connect methods
+    // in order to enable cardano derivation on a device
+    // https://github.com/trezor/trezor-firmware/blob/master/core/src/apps/cardano/README.md#seed-derivation-schemes
+    Object.keys(TrezorConnect)
+        .filter(k => !blacklist.includes(k as ConnectKey))
+        .forEach(key => {
+            // typescript complains about params and return type, need to be "any"
+            const original: any = TrezorConnect[key as ConnectKey];
```
Comment (juriczech) [inline suite-common/connect-init/src/cardanoConnectPatch.ts:44]: I understand that this is just copy paste to move the files around a little bit. But can we get rid of `any`? 

Nit: original is not really saying much about the variable. It looks weird to call asynchronous function with the name original.

File: suite-native/app/src/App.tsx [hunk 20]
```diff
@@ -1,60 +1,51 @@
...
+
+export const App = () => (
+    <GestureHandlerRootView style={{ flex: 1 }}>
+        <NavigationContainer>
+            <Provider store={store}>
+                <AppComponent />
+            </Provider>
+        </NavigationContainer>
+    </GestureHandlerRootView>
+);
```
Comment (juriczech) [inline suite-native/app/src/App.tsx]: Since we're separating the providers, wouldn't it make sense to have a component where we have all the providers and that will accept children, where the App would be passed? Now you have to look in two places to see which providers are wrapping the app.

**Review by juriczech — DISMISSED** [review]

File: suite-native/app/src/App.tsx (hunk 20 above)
Comment (bouchja1) [reply suite-native/app/src/App.tsx]: https://github.com/trezor/trezor-suite/pull/5760/commits/988384cfeee070ee525bbdbd42ef6d3a4d437999

File: suite-common/connect-init/src/actions/prepareConnectInit.ts (hunk 18 above)
Comment (bouchja1) [reply suite-common/connect-init/src/actions/prepareConnectInit.ts]: I am not sure if we can use it this way because it complains about types.

Snímek obrazovky 2022-07-13 v 14 29 22

the action in **suite** package is defined as:

`export const lockDevice = (payload: boolean): SuiteAction => ({
    type: SUITE.LOCK_DEVICE,
    payload,
});
`
it takes type for SuiteAction which is defined in suite actions type:

`
export type SuiteAction =
.....
| { type: typeof SUITE.LOCK_DEVICE; payload: boolean }
....
`

Comment (juriczech) [reply suite-common/connect-init/src/actions/prepareConnectInit.ts]: I think the biggest issue is that in App.tsx that calls for actions that do not exist and typescript doesn't even complain. It would require import of `SuiteAction` type to resolve this but that would create a circular dependency. 

```
 dispatch(
            prepareConnectInit({
                lockDevice: () => ({ type: 'NOOP' }),
                setInitConnectError: () => ({ type: 'NOOP' }),
                onConnectInitialized: () => ({ type: 'NOOP' }),
                enabledNetworks: [],
                initSettings: {
                    debug: true,
                    manifest: {
                        email: 'info@trezor.io',
                        appUrl: '@trezor/suite-native',
                    },
                },
            }),
        );
```
(`{ type: 'NOOP' }` doesn't exist and typescript doesn't even complain since the type is basically `any`)

I am curious whether we should move these `SuiteAction` types to some package which will export them and that would avoid circular dependency and achieve the result of having strictly typed actions and typescript complaining when dispatching action that doesn't exist. 
@Nodonisko @bouchja1

File: suite-common/connect-init/src/actions/prepareConnectInit.ts [hunk 21]
```diff
@@ -0,0 +1,46 @@
+import TrezorConnect, { ConnectSettings, Manifest } from '@trezor/connect';
```
Comment (juriczech) [inline suite-common/connect-init/src/actions/prepareConnectInit.ts]: ESLint complains about import order. `yarn lint --fix` should fix this.

File: suite-common/connect-init/src/cardanoConnectPatch.ts (hunk 19 above)
Comment (juriczech) [reply suite-common/connect-init/src/cardanoConnectPatch.ts:44]: Same in `App.tsx`

File: suite-common/connect-init/src/actions/prepareConnectInit.ts (hunk 21 above)
Comment (juriczech) [reply suite-common/connect-init/src/actions/prepareConnectInit.ts]: Same in `App.tsx`

File: suite-common/connect-init/src/actions/prepareConnectInit.ts (hunk 18 above)
Comment (Nodonisko) [reply suite-common/connect-init/src/actions/prepareConnectInit.ts]: We are not not going to use type for actions, typing actions is not necessary. `SuiteAction` is useless type, because it's is not going to guard you from using action that is doing  "nothing".

Currently we are defining our actions creators somewhere and then go and add that `typeof MyActionCreator`  to some global type `SuiteAction`. But if someone decide to delete this switch case from reducer so action is not handled anymore, it's still valid because someone manually placed it to Action type. So manually defining this type gives you exactly zero guarantee that actions is handled because this `SuiteAction` type is constructed manually anyway.

In case you will notify that action is not used anymore you will manually remove action creator and type so it will throw error, but it will throw error also in case there will be only action creator without strongly typed dispatch because you can't call non-existing function that was removed.

This is little bit better with redux-toolkit `createSlice` which will generate type automatically, but it also generate action creators automatically so if you delete that case, TS will throw anyway because action creator is not created anymore.

So if you are going to always use redux-toolkit slices for generating action creators and not constructing action object manually (`{ type: "Whatever", payload... }`) we will have even better type safety that with `SuiteAction` type.

`type: "NOOP"` is special case and I would prefer to use at least `createAction` for it.

File: suite-common/connect-init/src/actions/prepareConnectInit.ts [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+import { cardanoConnectPatch } from '../cardanoConnectPatch';
+import { lockDeviceConnectPatch } from '../lockDeviceConnectPatch';
+import { bindConnectActionsListeners, ConnectEventAction } from '../bindConnectActionsListeners';
+
+export const prepareConnectInit = createAsyncThunk<
+    void,
+    {
+        lockDevice: (payload: boolean) => AnyAction;
+        setInitConnectError: (payload: string) => AnyAction;
+        onConnectInitialized: () => AnyAction;
```
Comment (Nodonisko) [inline suite-common/connect-init/src/actions/prepareConnectInit.ts]: please use selectors `selectors` and `actions` as we talked before

File: suite-common/connect-init/src/cardanoConnectPatch.ts (hunk 19 above)
Comment (Nodonisko) [reply suite-common/connect-init/src/cardanoConnectPatch.ts:44]: I would prefer not mess with this, this is still quite small scoped PR but in future we will migrate bigger and more complicated pieces of code and doing to many changes together will be problem even if it's only renaming. We should do only very small refactorings especially on pieces of code that are not tested like this.

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/connect-init/src/actions/prepareConnectInit.ts (hunk 22 above)
Comment (bouchja1) [reply suite-common/connect-init/src/actions/prepareConnectInit.ts]: https://github.com/trezor/trezor-suite/pull/5760/commits/6e0c986bbb285014c511c1560640c419e0a2b7dd

File: suite-common/connect-init/src/actions/prepareConnectInit.ts (hunk 21 above)
Comment (bouchja1) [reply suite-common/connect-init/src/actions/prepareConnectInit.ts]: https://github.com/trezor/trezor-suite/pull/5760/commits/6e0c986bbb285014c511c1560640c419e0a2b7dd

File: suite-common/connect-init/src/actions/prepareConnectInit.ts (hunk 18 above)
Comment (bouchja1) [reply suite-common/connect-init/src/actions/prepareConnectInit.ts]: for now, solved with types:

```
type Selector<TReturnValue> = (state: any) => TReturnValue;

type SuiteCompatibleAction<TPayload = void> = (
    payload: TPayload,
) => AnyAction | ActionCreatorWithPayload<TPayload> | ActionCreatorWithoutPayload;
```

File: suite-common/connect-init/src/actions/prepareConnectInit.ts [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+import { bindConnectActionsListeners, ConnectEventAction } from '../bindConnectActionsListeners';
+
+type Selector<TReturnValue> = (state: any) => TReturnValue;
+type SuiteCompatibleAction<TPayload = void> = (
+    payload: TPayload,
+) => AnyAction | ActionCreatorWithPayload<TPayload> | ActionCreatorWithoutPayload;
+
+export const CONNECT_INIT_ACTION_TYPE = '@suite-common/connect-init/init';
+
+export const prepareConnectInit = createAsyncThunk<
```
Comment (Nodonisko) [inline suite-common/connect-init/src/actions/prepareConnectInit.ts]: We should start suffixing all thunk with thunk. `prepareConnectInit` => `prepareConnectInitThunk`

File: suite-common/connect-init/src/lockDeviceConnectPatch.ts [hunk 24]
```diff
@@ -0,0 +1,39 @@
+import TrezorConnect from '@trezor/connect';
```
Comment (Nodonisko) [inline suite-common/connect-init/src/lockDeviceConnectPatch.ts]: Can you move this stuff back to `prepareConnectInit` and maybe even get rid of `lockDeviceConnectPatch ` function and use it's code directly in thunk (with dispatches instead of callbacks etc.)

File: suite-common/connect-init/src/actions/prepareConnectInit.ts [hunk 25]
```diff
@@ -0,0 +1,70 @@
+import {
```
Comment (Nodonisko) [inline suite-common/connect-init/src/actions/prepareConnectInit.ts]: Can we rename this file `slice.ts` and move it to root, to follow conventions?

File: suite-common/connect-init/src/bindConnectActionsListeners.ts [hunk 26]
```diff
@@ -0,0 +1,39 @@
+import TrezorConnect, {
```
Comment (Nodonisko) [inline suite-common/connect-init/src/bindConnectActionsListeners.ts]: Can you move this back to thunk and get rid of functions, eg use directly dispatches.

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/connect-init/src/actions/prepareConnectInit.ts (hunk 23 above)
Comment (bouchja1) [reply suite-common/connect-init/src/actions/prepareConnectInit.ts]: https://github.com/trezor/trezor-suite/pull/5760/commits/979f44f2e30665c851548bc75ae31ae9d9bbba9b

File: suite-common/connect-init/src/lockDeviceConnectPatch.ts (hunk 24 above)
Comment (bouchja1) [reply suite-common/connect-init/src/lockDeviceConnectPatch.ts]: https://github.com/trezor/trezor-suite/pull/5760/commits/979f44f2e30665c851548bc75ae31ae9d9bbba9b

File: suite-common/connect-init/src/actions/prepareConnectInit.ts (hunk 25 above)
Comment (bouchja1) [reply suite-common/connect-init/src/actions/prepareConnectInit.ts]: https://github.com/trezor/trezor-suite/pull/5760/commits/979f44f2e30665c851548bc75ae31ae9d9bbba9b

File: suite-common/connect-init/src/bindConnectActionsListeners.ts (hunk 26 above)
Comment (bouchja1) [reply suite-common/connect-init/src/bindConnectActionsListeners.ts]: https://github.com/trezor/trezor-suite/pull/5760/commits/979f44f2e30665c851548bc75ae31ae9d9bbba9b

File: packages/suite/src/actions/suite/trezorConnectActions.ts [hunk 27]
```diff
@@ -1,146 +1,35 @@
...
-};
+export const init = prepareConnectInitThunk({
+    actions: {
+        lockDevice,
+    },
+    selectors: {
+        selectEnabledNetworks,
+        selectIsPendingTransportEvent,
+    },
+    initSettings: {
```
Comment (Nodonisko) [inline packages/suite/src/actions/suite/trezorConnectActions.ts]: nit: no need to spread, just `initSettings,` will work too

File: suite-common/connect-init/redux.d.ts [hunk 28]
```diff
@@ -0,0 +1,7 @@
+import { AsyncThunkAction } from '@reduxjs/toolkit';
```
Comment (Nodonisko) [inline suite-common/connect-init/redux.d.ts:1]: is this file necessary?

File: suite-native/state/src/configureStore.ts [hunk 29]
```diff
@@ -0,0 +1,17 @@
+import { configureStore, AsyncThunkAction } from '@reduxjs/toolkit';
```
Comment (Nodonisko) [inline suite-native/state/src/configureStore.ts]: Why is this file here? We already have store initialized in our state package

File: suite-native/app/src/App.tsx [hunk 30]
```diff
@@ -38,18 +34,36 @@ export const App = () => {
...
-            })
-            .catch(error => {
-                // eslint-disable-next-line no-console
-                console.log('Init failed', JSON.stringify(error.code));
-            });
-    }, [getAccountInfo]);
+        const noOperation = createAction('noOperation');
+
+        // TODO handle possible error
+        dispatch(
```
Comment (Nodonisko) [inline suite-native/app/src/App.tsx]: please move this prepare before component
```
const initConnectThunk = prepareConnectInitThunk({ })
// later in component
dispatch(initConnectThunk)
```

File: suite-common/connect-init/src/__tests__/trezorConnectActions.test.ts [hunk 31]
```diff
@@ -0,0 +1,217 @@
...
+
+const init = prepareConnectInitThunk({
+    actions: {
+        lockDevice: createAction<boolean>(MOCK_LOCK_DEVICE_ACTION),
+    },
+    selectors: {
+        selectEnabledNetworks: () => [],
+        selectIsPendingTransportEvent: () => true,
+    },
+    initSettings: {
```
Comment (Nodonisko) [inline suite-common/connect-init/src/__tests__/trezorConnectActions.test.ts]: no need to spread

File: suite-common/connect-init/src/__tests__/trezorConnectActions.test.ts [hunk 32]
```diff
@@ -0,0 +1,217 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
```
Comment (Nodonisko) [inline suite-common/connect-init/src/__tests__/trezorConnectActions.test.ts:1]: Please remove original test file in the end it doesn't make sense to have it twice

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/connect-init/redux.d.ts (hunk 28 above)
Comment (bouchja1) [reply suite-common/connect-init/redux.d.ts:1]: It is necessary. Otherwise it is failing on type-check:

Snímek obrazovky 2022-08-04 v 10 59 21

File: suite-common/connect-init/src/__tests__/trezorConnectActions.test.ts (hunk 32 above)
Comment (bouchja1) [reply suite-common/connect-init/src/__tests__/trezorConnectActions.test.ts:1]: I am not sure. There will probably be a difference in error handling - when an error occurs in `prepareConnectInit`, **suite** is catching the error and dispatch further action to deal with it (that's the reason I have added `.unwrap() ` here https://github.com/trezor/trezor-suite/pull/5760/files#diff-2acc8432632da4d879e822e9644237af62d547114f0f092524383540492071f3R47). While `prepareConnectInit` is async thunk from redux/toolkit, **suite-native** will handle _prepareConnectInit.rejected_ thunk state. And this is the difference in both tests and I would prefer not to unify them at least for now.

**Review by Nodonisko — APPROVED** [review]
